File: 2021/08_SevenSegmentSearch/patterns.py
# ======================================================================
# Seven Segment Search
#   Advent of Code 2021 Day 08 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                         p a t t e r n s . p y
# ======================================================================
"A solver for the Advent of Code 2021 Day 08 puzzle"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------
import pattern

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------

# ======================================================================
#                                                               Patterns
# ======================================================================


class Patterns(object):   # pylint: disable=R0902, R0205
    "Object for Seven Segment Search"

    def __init__(self, text=None, part2=False):

        # 1. Set the initial values
        self.part2 = part2
        self.text = text
        self.patterns = []

        # 2. Process text (if any)
        if text is not None and len(text) > 0:
            for line in text:
                self.patterns.append(pattern.Pattern(text=line, part2=part2))

    def one_four_seven_eight(self):
        "Return the number of those values"

        # 1. Start with nothing
        result = 0

        # 2. Loop for the values
        for ptrn in self.patterns:

            # 3. Increment the result by the number of desired digits
            result += ptrn.one_four_seven_eight()

        # 4. Return the count of the desired digits
        return result

    def total_values(self):
        "Return the number of those values"

        # 1. Start with nothing
        result = 0

        # 2. Loop for the values
        for ptrn in self.patterns:

            # 3. Increment the result by the pattern's value
            result += ptrn.value()

        # 4. Return the sum of the values
        return result

    def part_one(self, verbose=False, limit=0):
        "Returns the solution for part one"

        # 0. Precondition axioms
        assert verbose in [True, False]
        assert limit >= 0

        # 1. Return the solution for part one
        return self.one_four_seven_eight()

    def part_two(self, verbose=False, limit=0):
        "Returns the solution for part two"

        # 0. Precondition axioms
        assert verbose in [True, False]
        assert limit >= 0

        # 1. Return the solution for part two
        return self.total_values()


# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------
if __name__ == '__main__':
    pass

# ======================================================================
# end                      p a t t e r n s . p y                     end
# ======================================================================

File: 2023/13_PointofIncidence/patterns.go
// ======================================================================
// Point Of Incidence
//   Advent of Code 2023 Day 13 -- Eric Wastl -- https://adventofcode.com
//
// go implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                         p a t t e r n s . g o
// ======================================================================
// A solver for the Advent of Code 2023 Day 13 puzzle

package main

// ----------------------------------------------------------------------
//                                                                imports
// ----------------------------------------------------------------------
import "fmt"

// ----------------------------------------------------------------------
//
//	constants
//
// ----------------------------------------------------------------------
const ()

// ======================================================================
//                                                               Patterns
// ======================================================================

// Patterns ... Object for Point Of Incidence
type Patterns struct {
	Part2    bool
	Text     []string
	Patterns []*Pattern
}

// NewPatterns ... Create a Patterns object
func NewPatterns(part2 bool, text []string) (*Patterns, error) {

	// 1. Set the initial values
	p := new(Patterns)
	p.Part2 = part2
	p.Text = text
	p.Patterns = []*Pattern{}

	// 2. Process text (if any)
	if len(p.Text) > 0 {
		err := p.processText(p.Text)
		if err != nil {
			return nil, err
		}
	}

	// 3. Return the new Patterns object
	return p, nil
}

// processText ... Assign values from text
func (p *Patterns) processText(text []string) error {

	// 1. Start with an empty pattern
	pattern, _ := NewPattern(p.Part2, "")

	// 1. Loop for each line of the text
	for _, line := range text {

		// 2. Add the text to the current pattern
		if line == "" {
			if pattern.HasText() {
				pattern.Rotate()
				p.Patterns = append(p.Patterns, pattern)
				pattern, _ = NewPattern(p.Part2, "")
			}
		} else {
			pattern.AddText(line)
		}
	}
	if pattern.HasText() {
		pattern.Rotate()
		p.Patterns = append(p.Patterns, pattern)
	}

	// 3. Return success
	return nil
}

// Score ... Return the total of the scores of all the patterns
func (p *Patterns) Score() int {

	// 1. Start with nothing
	result := 0

	// 2. Loop for all the patterns
	for _, pattern := range p.Patterns {

		// 3. Add in the score for this pattern
		result = result + pattern.Score()
	}

	// 4. Return the total of the scores
	return result
}

// PartOne ... Returns the solution for part one
func (p *Patterns) PartOne(verbose bool, limit int) string {

	// 1. Return the solution for part one
	return fmt.Sprint(p.Score())
}

// PartTwo ... Returns the solution for part two
func (p *Patterns) PartTwo(verbose bool, limit int) string {

	// 1. Return the solution for part two
	return fmt.Sprint(p.Score())
}

// ======================================================================
// end                      p a t t e r n s . g o                     end
// ======================================================================
